File: client/source/Proton.Stream.cs
using System.Collections;
using System.Collections.Generic;
using System;
using System.Text;
using UnityEngine;
using Proton.Packet.Serialization;
using Proton.Packet.ID;

namespace Proton.Stream
{
    public class ProtonStream
    {
        public List<byte> Bytes = new List<byte>();
        public short ReadOffset = 0;

        public void WriteByte(byte value)
        {
            byte castedValue = (byte) value;
            Bytes.Add(castedValue);
        }
        public void WriteBytes(byte[] values)
        {
            foreach (byte value in (byte[]) values)
            {
                WriteByte(value);
            }
        }
        public void WriteInt16(short value)
        {
            short castedValue = (short) value;
            byte[] shortBytes = BitConverter.GetBytes(castedValue);
            for (int i = 0; i < shortBytes.Length; i++)
            {
                WriteByte(shortBytes[i]);
            }
        }
        public void WriteUInt16(ushort value)
        {
            ushort castedValue = (ushort) value;
            byte[] ushortBytes = BitConverter.GetBytes(castedValue);
            for (int i = 0; i < ushortBytes.Length; i++)
            {
                WriteByte(ushortBytes[i]);
            }
        }
        public void WriteInt32(int value)
        {
            int castedValue = (int) value;
            byte[] longBytes = BitConverter.GetBytes(castedValue);
            for (int i = 0; i < longBytes.Length; i++)
            {
                WriteByte(longBytes[i]);
            }
        }
        public void WriteUInt32(uint value)
        {
            uint castedValue = (uint) value;
            byte[] ulongBytes = BitConverter.GetBytes(castedValue);
            for (int i = 0; i < ulongBytes.Length; i++)
            {
                WriteByte(ulongBytes[i]);
            }
        }
        public void WriteFloat(float value)
        {
            byte[] floatBytes = BitConverter.GetBytes(value);
            for (int i = 0; i < floatBytes.Length; i++)
            {
                WriteByte(floatBytes[i]);
            }
        }
        public void WriteString(string value)
        {
            byte[] stringBytes = Encoding.UTF8.GetBytes(value);
            ushort length = (ushort) stringBytes.Length;

            WriteUInt16(length);
            for (int i = 0; i < length; i++)
            {
                WriteByte(stringBytes[i]);
            }
        }
        public void WriteBool(bool value)
        {
            WriteByte((byte)  (value ? 1 : 0));
        }
        public void WriteVector3(Vector3 value)
        {
            WriteFloat(value.x);
            WriteFloat(value.y);
            WriteFloat(value.z);
        }
        public void WriteQuaternion(Quaternion value)
        {
            WriteFloat(value[0]);
            WriteFloat(value[1]);
            WriteFloat(value[2]);
            WriteFloat(value[3]);
        }
        public void WriteBytearray(byte[] array)
        {
            WriteUInt32((uint) array.Length);
            WriteBytes(array);
        }

        public T Read<T>()
        {
            if (typeof(T) == typeof(bool))
            {
                return (T)(object)ReadBool();
            }
            else if (typeof(T) == typeof(byte))
            {
                return (T)(object)ReadByte();
            }
            else if (typeof(T) == typeof(ushort))
            {
                return (T)(object)ReadUInt16();
            }
            else if (typeof(T) == typeof(short))
            {
                return (T)(object)ReadInt16();
            }
            else if (typeof(T) == typeof(uint))
            {
                return (T)(object)ReadUInt32();
            }
            else if (typeof(T) == typeof(int))
            {
                return (T)(object)ReadInt32();
            }
            else if (typeof(T) == typeof(float))
            {
                return (T)(object)ReadFloat();
            }
            else if (typeof(T) == typeof(string))
            {
                return (T)(object)ReadString();
            }
            else if (typeof(T) == typeof(Vector3))
            {
                return (T)(object)ReadVector3();
            }
            else if (typeof(T) == typeof(Quaternion))
            {
                return (T)(object)ReadQuaternion();
            }
            else if (typeof(T) == typeof(byte[]))
            {
                return (T)(object)ReadBytearray();
            }
            else
            {
                throw new InvalidOperationException("Invalid type provided for reading from ProtonStream.");
            }
        }
        public byte ReadByte()
        {
            ReadOffset++;
            return Bytes[ReadOffset - 1];
        }
        public byte[] ReadBytes(uint ammount)
        {
            List<byte> result = new List<byte>();
            for (int i = 0; i < (int) ammount; i++)
            {
                result.Add(ReadByte());
            }
            return result.ToArray();
        }
        public short ReadInt16()
        {
            List<byte> shortBytes = new List<byte>();
            for (int i = 0; i < 2; i++)
            {
                shortBytes.Add(ReadByte());
            }
            return BitConverter.ToInt16(shortBytes.ToArray(), 0);
        }
        public ushort ReadUInt16()
        {
            List<byte> ushortBytes = new List<byte>();
            for (int i = 0; i < 2; i++)
            {
                ushortBytes.Add(ReadByte());
            }
            return BitConverter.ToUInt16(ushortBytes.ToArray(), 0);
        }
        public int ReadInt32()
        {
            List<byte> intBytes = new List<byte>();
            for (int i = 0; i < 4; i++)
            {
                intBytes.Add(ReadByte());
            }
            return BitConverter.ToInt32(intBytes.ToArray(), 0);
        }
        public uint ReadUInt32()
        {
            List<byte> uintBytes = new List<byte>();
            for (int i = 0; i < 4; i++)
            {
                uintBytes.Add(ReadByte());
            }
            return BitConverter.ToUInt32(uintBytes.ToArray(), 0);
        }
        public float ReadFloat()
        {
            List<byte> floatBytes = new List<byte>();
            for (int i = 0; i < 4; i++)
            {
                floatBytes.Add(ReadByte());
            }
            return BitConverter.ToSingle(floatBytes.ToArray(), 0);
        }
        public string ReadString()
        {
            ushort length = ReadUInt16();
            List<byte> stringBytes = new List<byte>();

            for (int i = 0; i < length; i++)
            {
                stringBytes.Add(ReadByte());
            }
    
            return Encoding.UTF8.GetString(stringBytes.ToArray());
        }
        public bool ReadBool()
        {
            return ReadByte() == 1;
        }
        public Vector3 ReadVector3()
        {
            return new Vector3(ReadFloat(), ReadFloat(), ReadFloat());
        }
        public Quaternion ReadQuaternion()
        {
            Quaternion quaternion = new Quaternion();
            quaternion[0] = ReadFloat();
            quaternion[1] = ReadFloat();
            quaternion[2] = ReadFloat();
            quaternion[3] = ReadFloat();
            return quaternion;
        }
        public byte[] ReadBytearray()
        {
            uint length = ReadUInt32();
            return ReadBytes(length);
        }

        public void WriteNetworkValue(NetworkValue networkValue)
        {
            byte valueType = networkValue.type;
            object value = networkValue.value;

            WriteByte(valueType);
            if (valueType == ProtonPacketID.BOOL)
            {
                WriteBool((bool) value);
            }
            else if (valueType == ProtonPacketID.BYTE)
            {
                WriteByte((byte) value);
            }
            else if (valueType == ProtonPacketID.UINT16)
            {
                WriteUInt16((ushort) value);
            }
            else if (valueType == ProtonPacketID.INT16)
            {
                WriteInt16((short) value);
            }
            else if (valueType == ProtonPacketID.UINT32)
            {
                WriteUInt32((uint) value);
            }
            else if (valueType == ProtonPacketID.INT32)
            {
                WriteInt32((int) value);
            }
            else if (valueType == ProtonPacketID.FLOAT)
            {
                WriteFloat((float) value);
            }
            else if (valueType == ProtonPacketID.STRING)
            {
                WriteString((string) value);
            }
            else if (valueType == ProtonPacketID.VECTOR3)
            {
                WriteVector3((Vector3) value);
            }
            else if (valueType == ProtonPacketID.QUATERNION)
            {
                WriteQuaternion((Quaternion) value);
            }
            else if (valueType == ProtonPacketID.BYTEARRAY)
            {
                WriteBytearray((byte[]) value);
            }
        }
        public NetworkValue ReadNetworkValue()
        {
            byte valueType = ReadByte();
            object value = null;
            
            if (valueType == ProtonPacketID.BOOL)
            {
                value = Read<bool>();
            }
            else if (valueType == ProtonPacketID.BYTE)
            {
                value = Read<byte>();
            }
            else if (valueType == ProtonPacketID.UINT16)
            {
                value = Read<ushort>();
            }
            else if (valueType == ProtonPacketID.INT16)
            {
                value = Read<short>();
            }
            else if (valueType == ProtonPacketID.UINT32)
            {
                value = Read<uint>();
            }
            else if (valueType == ProtonPacketID.INT32)
            {
                value = Read<int>();
            }
            else if (valueType == ProtonPacketID.FLOAT)
            {
                value = Read<float>();
            }
            else if (valueType == ProtonPacketID.STRING)
            {
                value = Read<string>();
            }
            else if (valueType == ProtonPacketID.VECTOR3)
            {
                value = Read<Vector3>();
            }
            else if (valueType == ProtonPacketID.QUATERNION)
            {
                value = Read<Quaternion>();
            }
            else if (valueType == ProtonPacketID.BYTEARRAY)
            {
                value = Read<byte[]>();
            }

            return new NetworkValue(valueType, value);
        }

        public void WriteNetworkDictionary(NetworkDictionary networkDictionary)
        {
            Dictionary<string, NetworkValue> dictionary = networkDictionary.dictionary;

            WriteUInt16((ushort) dictionary.Count);
            foreach (string key in dictionary.Keys)
            {
                WriteString(key);
                WriteNetworkValue(dictionary[key]);
            }
        }
        public NetworkDictionary ReadNetworkDictionary()
        {
            NetworkDictionary resultDictionary = new NetworkDictionary();

            ushort valuesAmmount = ReadUInt16();
            for (int i = 0; i < (int) valuesAmmount; i++)
            {
                resultDictionary.dictionary[ReadString()] = ReadNetworkValue();
            }

            return resultDictionary;
        }
    }
}

File: server/source/protonstream.py
import struct
from utils import constrain
from enums import *
from networkvalue import *
from networkdictionary import *
from customtypes import *

class ProtonStream:
    def __init__(self):
        self.bytes = []
        self.readOffset = 0

    def setBytes(self, bytes):
        self.bytes = list(bytes)
        self.readOffset = 0

    def getBytes(self):
        return bytes(self.bytes)

    def writeByte(self, value):
        value = value & 0xFF
        self.bytes.append(value)

    def writeBytes(self, values):
        values = list(values)
        for value in values:
            self.writeByte(value)

    def writeInt16(self, value):
        value = constrain(value, -0x7FFF, 0x7FFF)
        shortBytes = (value).to_bytes(2, byteorder="little", signed=True)
        for i in range(2):
            self.writeByte(shortBytes[i])

    def writeUInt16(self, value):
        value = constrain(value, 0, 0xFFFF)
        shortBytes = (value).to_bytes(2, byteorder="little", signed=False)
        for i in range(2):
            self.writeByte(shortBytes[i])

    def writeInt32(self, value):
        value = constrain(value, -0x7FFFFFFF, 0x7FFFFFFF)
        longBytes = (value).to_bytes(4, byteorder="little", signed=True)
        for i in range(4):
            self.writeByte(longBytes[i])

    def writeUInt32(self, value):
        value = constrain(value, 0, 0xFFFFFFFF)
        longBytes = (value).to_bytes(4, byteorder="little", signed=False)
        for i in range(4):
            self.writeByte(longBytes[i])

    def writeFloat(self, value):
        floatBytes = bytes(struct.pack("f", value))
        for i in range(4):
            self.writeByte(floatBytes[i])

    def writeString(self, value):
        stringBytes = value.encode("utf-8", "replace")

        stringLength = len(stringBytes)
        if stringLength > 0xFFFF:
            stringLength = 0xFFFF

        self.writeUInt16(stringLength)
        for i in range(stringLength):
            self.writeByte(stringBytes[i])

    def writeBool(self, value):
        self.writeByte(1 if value else 0)

    def writeBytearray(self, values):
        values = list(values)
        self.writeUInt32(len(values))
        self.writeBytes(values)

    def writeVector3(self, value):
        self.writeFloat(value.x)
        self.writeFloat(value.y)
        self.writeFloat(value.z)

    def writeVector2(self, value):
        self.writeFloat(value.x)
        self.writeFloat(value.y)

    def writeQuaternion(self, value):
        self.writeFloat(value.x)
        self.writeFloat(value.y)
        self.writeFloat(value.z)
        self.writeFloat(value.w)

    def readByte(self):
        self.readOffset += 1
        return self.bytes[self.readOffset - 1]

    def readBytes(self, ammount):
        readedBytes = []
        for i in range(ammount):
            readedBytes.append(self.readByte())
        return readedBytes

    def readInt16(self):
        shortBytes = []
        for i in range(2):
            shortBytes.append(self.readByte())
        return int.from_bytes(shortBytes, byteorder="little", signed=True)

    def readUInt16(self):
        shortBytes = []
        for i in range(2):
            shortBytes.append(self.readByte())
        return int.from_bytes(shortBytes, byteorder="little", signed=False)

    def readInt32(self):
        intBytes = []
        for i in range(4):
            intBytes.append(self.readByte())
        return int.from_bytes(intBytes, byteorder="little", signed=True)

    def readUInt32(self):
        intBytes = []
        for i in range(4):
            intBytes.append(self.readByte())
        return int.from_bytes(intBytes, byteorder="little", signed=False)

    def readFloat(self):
        floatBytes = []
        for i in range(4):
            floatBytes.append(self.readByte())
        return struct.unpack("f", bytes(floatBytes))[0]

    def readString(self):
        stringLength = self.readUInt16()
        stringBytes = []
        for i in range(stringLength):
            stringBytes.append(self.readByte())
        return bytes(stringBytes).decode("utf-8")

    def readBool(self):
        return self.readByte() == 1

    def readBytearray(self):
        length = self.readUInt32()
        return list(self.readBytes(length))

    def readVector3(self):
        return Vector3(self.readFloat(), self.readFloat(), self.readFloat())

    def readVector2(self):
        return Vector2(self.readFloat(), self.readFloat())

    def readQuaternion(self):
        return Quaternion(self.readFloat(), self.readFloat(), self.readFloat())

    def writeNetworkValue(self, networkValue):
        valueType = networkValue.type
        value = networkValue.value
        
        self.writeByte(valueType)
        if valueType == BOOL:
            self.writeBool(value)
        elif valueType == BYTE:
            self.writeByte(value)
        elif valueType == UINT16:
            self.writeUInt16(value)
        elif valueType == INT16:
            self.writeInt16(value)
        elif valueType == UINT32:
            self.writeUInt32(value)
        elif valueType == INT32:
            self.writeInt32(value)
        elif valueType == FLOAT:
            self.writeFloat(value)
        elif valueType == STRING:
            self.writeString(value)
        elif valueType == BYTEARRAY:
            self.writeBytearray(value)
        elif valueType == VECTOR3:
            self.writeVector3(value)
        elif valueType == VECTOR2:
            self.writeVector2(value)
        elif valueType == QUATERNION:
            self.writeQuaternion(value)

    def readNetworkValue(self):
        valueType = self.readByte()
        value = None

        if valueType == BOOL:
            value = self.readBool()
        elif valueType == BYTE:
            value = self.readByte()
        elif valueType == UINT16:
            value = self.readUInt16()
        elif valueType == INT16:
            value = self.readInt16()
        elif valueType == UINT32:
            value = self.readUInt32()
        elif valueType == INT32:
            value = self.readInt32()
        elif valueType == FLOAT:
            value = self.readFloat()
        elif valueType == STRING:
            value = self.readString()
        elif valueType == BYTEARRAY:
            value = self.readBytearray()
        elif valueType == VECTOR3:
            value = self.readVector3()
        elif valueType == VECTOR2:
            value = self.readVector2()
        elif valueType == QUATERNION:
            value = self.readQuaternion()

        return NetworkValue(valueType, value)

    def writeNetworkDictionary(self, networkDictionary):
        dictionary = networkDictionary.dictionary
        self.writeUInt16(len(dictionary))
        for key in dictionary:
            self.writeString(key)
            self.writeNetworkValue(dictionary[key])

    def readNetworkDictionary(self):
        dictionarySize = self.readUInt16()
        dictionary = {}
        for i in range(dictionarySize):
            key = self.readString()
            dictionary[key] = self.readNetworkValue()

        return NetworkDictionary(dictionary)